README.md
# StatusCake-Helpers
[![Build status](https://dev.azure.com/oliverli0875/Helpers/_apis/build/status/Oliver-Lii.statuscake-helpers?branchName=master)](https://dev.azure.com/oliverli0875/Helpers/_build?definitionId=1&_a=summary&repositoryFilter=1&branchFilter=2) [![Test Results](https://img.shields.io/azure-devops/tests/oliverli0875/Helpers/1.svg)](https://dev.azure.com/oliverli0875/Helpers/_build?definitionId=1&_a=summary&view=ms.vss-pipelineanalytics-web.new-build-definition-pipeline-analytics-view-cardmetrics) [![GitHub license](https://img.shields.io/github/license/Oliver-Lii/StatusCake-Helpers.svg)](LICENSE) [![PowerShell Gallery](https://img.shields.io/powershellgallery/v/StatusCake-Helpers.svg)](https://www.powershellgallery.com/packages/statuscake-helpers/)


This module was written to support interaction with the Statuscake API via Powershell. Additional functionality may be added later and I will use this as a generic module to house Powershell functions specific to interacting with the Statuscake API.

**DISCLAIMER:** Neither this module, nor its creator are in any way affiliated with StatusCake.


# Usage
This module can be installed from the PowerShell Gallery using the command below.
```powershell
Install-Module StatusCake-Helpers -Repository PSGallery
```

## Example

 The following illustrates how to create uptime, SSL, Page Speed Tests and a Public Reporting page along with daily and weekly maintenance windows and two contacts

```powershell
# Setup the StatusCake API key
# The API credentials must come from the primary account which hosts the tests and not a subaccount which was given access
$scAPIKey = Read-Host -AsSecureString -Prompt "Please enter the API key"
Set-StatusCakeHelperAPIAuth -APIKey $scAPIKey

$URL = "https://www.example.com"
$team1Emails = @("alerts@example.com","alerts1@example.com")

$team1Contact= New-StatusCakeHelperContactGroup -Name "Team 1 monitoring" -Email $team1Emails -Mobile "+14155552671" -Passthru
$team2Contact= New-StatusCakeHelperContactGroup -Name "Team 2 monitoring" -Email "alerts2@example.com" -Passthru

#Create a uptime test to check the site every 5 minutes
$uptimeTest = New-StatusCakeHelperUptimeTest -Name "Example" -TestURL $URL -CheckRate 300 -TestType HTTP -ContactGroup $team1Contact.id

#Create SSL test to check SSL certificate every day
$sslTest = New-StatusCakeHelperSSLTest -Domain $URL -Checkrate 2073600 -ContactID @($team1Contact.id,$team2Contact.id)

#Create Page Speed Test to monitor page speed every 30 minutes from the UK
$pageSpeedCheckName = "Example site UK speed check"
$pageSpeedTest = New-StatusCakeHelperPageSpeedTest -Name $pageSpeedCheckName -WebsiteURL $URL -Checkrate 1800 -Region UK

#Update the page speed test using the name of the test to alert team 2 when the page takes more than 5000ms to load
$result = Update-StatusCakeHelperPageSpeedTest -Name $pageSpeedCheckName -ContactID @($team2Contact.id) -AlertSlower 5000

#Create a date object to start today at 20:00 and finish in an hour
$startMWDailyTime = Get-Date "20:00"
$endMWDailyTime = $startMWDailyTime.AddHours(1)

#Setup a date object to start next Saturday at 20:00 and finish in four hours time
$startMWWeeklyTime = $startMWDailyTime
while ($startMWWeeklyTime.DayOfWeek -ne "Saturday")
{
    $startMWWeeklyTime = $startMWWeeklyTime.AddDays(1)
}
$endMWWeeklyTime = $startMWWeeklyTime.AddHours(4)

$mwParams = @{
    Timezone = "Europe/London"
    UptimeID = @($uptimeTest.id)
}

#Create the daily reoccurring maintenance window
$result = New-StatusCakeHelperMaintenanceWindow -Name "Example Daily MW" -StartDate $startMWDailyTime -EndDate $endMWDailyTime @mwParams -RepeatInterval 1d

#Create the weekly reoccurring maintenance window
$result = New-StatusCakeHelperMaintenanceWindow -Name "Example Weekly MW" -StartDate $startMWWeeklyTime -EndDate $endMWWeeklyTime @mwParams -RepeatInterval 1w

```

## Functions

Below is a list of the available functions and features of the StatusCake API that are supported. Further details of each function can be found in the links below:

[Authentication](https://github.com/Oliver-Lii/statuscake-helpers/tree/master/Documentation/Authentication "StatusCake API Authentication")
*  Remove-StatusCakeHelperAPIAuth
*  Update-StatusCakeHelperAPIAuth
*  Test-StatusCakeHelperAPIAuthSet

[ContactGroups](https://github.com/Oliver-Lii/statuscake-helpers/tree/master/Documentation/ContactGroups "StatusCake Contact Groups")
*  Copy-StatusCakeHelperContactGroup
*  Get-StatusCakeHelperContactGroup
*  New-StatusCakeHelperContactGroup
*  Remove-StatusCakeHelperContactGroup
*  Update-StatusCakeHelperContactGroup

[Locations](https://github.com/Oliver-Lii/statuscake-helpers/tree/master/Documentation/Locations "StatusCake Locations")
*  Get-StatusCakeHelperLocation

[MaintenanceWindows](https://github.com/Oliver-Lii/statuscake-helpers/tree/master/Documentation/MaintenanceWindows "StatusCake Maintenance Windows")
*  Get-StatusCakeHelperMaintenanceWindow
*  New-StatusCakeHelperMaintenanceWindow
*  Remove-StatusCakeHelperMaintenanceWindow
*  Update-StatusCakeHelperMaintenanceWindow

[PageSpeed](https://github.com/Oliver-Lii/statuscake-helpers/tree/master/Documentation/PageSpeed "StatusCake PageSpeed Tests")
*  Copy-StatusCakeHelperPageSpeedTest
*  Get-StatusCakeHelperPageSpeedTest
*  Get-StatusCakeHelperPageSpeedTestHistory
*  New-StatusCakeHelperPageSpeedTest
*  Remove-StatusCakeHelperPageSpeedTest
*  Update-StatusCakeHelperPageSpeedTest

[SSL](https://github.com/Oliver-Lii/statuscake-helpers/tree/master/Documentation/SSL "StatusCake SSL Tests")
*  Copy-StatusCakeHelperSSLTest
*  Get-StatusCakeHelperSSLTest
*  New-StatusCakeHelperSSLTest
*  Remove-StatusCakeHelperSSLTest
*  Update-StatusCakeHelperSSLTest

[Uptime](https://github.com/Oliver-Lii/statuscake-helpers/tree/master/Documentation/Uptime "StatusCake Tests")
*  Copy-StatusCakeHelperUptimeTest
*  Get-StatusCakeHelperUptimeAlert
*  Get-StatusCakeHelperUptimeHistory
*  Get-StatusCakeHelperUptimePeriod
*  Get-StatusCakeHelperUptimeTest
*  Get-StatusCakeHelperUptimePausedTest
*  New-StatusCakeHelperUptimeTest
*  Remove-StatusCakeHelperUptimeTest
*  Resume-StatusCakeHelperUptimeTest
*  Update-StatusCakeHelperUptimeTest
*  Suspend-StatusCakeHelperUptimeTest

## Tests

This module comes with [Pester](https://github.com/pester/Pester/) tests for unit testing. The tests cover the StatusCake features available under the [Business plan](https://www.statuscake.com/pricing/).


# Authors
- Oliver Li

ReleaseNotes.md
# Release Notes

## Release 5.0

*  Public Reporting functions have removed as this is no longer supported by the API
*  Functions for retrieving StatusCake `probes` have been replaced by `location` functions to reflect change in StatusCake terminology
*  Test functions have been renamed to include the `Uptime` prefix to clarify they are for uptime tests
*  `Set` functions have been replaced by `Update` functions
*  Detail functions for PageSpeed and Uptime test removed. Test details are returned by default when a test is fetched by name or ID
*  Contact group and Uptime test name parameters have been renamed as `Name` instead of `GroupName` or `TestName`

### Documentation
*  Remove Public Reporting documentation as this is no longer supported by the API
*  Removed Alerts, PerformanceData and PeriodData folders and cmdlets as these commands have been superceded

## Summary of function changes

| Function name                             	| Status                                                                    	|
|-------------------------------------------	|---------------------------------------------------------------------------	|
| Get-StatusCakeHelperSentAlerts            	| Deprecated - functionality moved to Get-StatusCakeHelperUptimeAlert           |
| Get-StatusCakeHelperPerformanceData           | Deprecated - functionality moved to Get-StatusCakeHelperUptimeHistory         |
| Get-StatusCakeHelperProbe                 	| Deprecated - functionality moved to Get-StatusCakeHelperLocation             	|
| Get-StatusCakeHelperPeriodOfData              | Deprecated - functionality moved to Get-StatusCakeHelperUptimePeriod          |
| Get-StatusCakeHelperTestDetail                | Deprecated - functionality moved to Get-StatusCakeHelperUptimeTest            |
|                                               |                                                                               |
| Set-StatusCakeHelperContactGroup              | Deprecated - replaced by Update-StatusCakeHelperContactGroup                  |
| Set-StatusCakeHelperPageSpeedTest             | Deprecated - replaced by Update-StatusCakeHelperPageSpeedTest                 |
| Set-StatusCakeHelperSSLTest                   | Deprecated - replaced by Update-StatusCakeHelperSSLTest                       |
| Set-StatusCakeHelperUptimeTest                | Deprecated - replaced by Update-StatusCakeHelperUptimeTest                    |
|                                           	|                                                                           	|
| Copy-StatusCakeHelperPublicReportingPage     	| Removed no longer supported by API                                            |
| Get-StatusCakeHelperPublicReportingPage      	| Removed no longer supported by API                                            |
| Get-StatusCakeHelperPublicReportingPageDetail	| Removed no longer supported by API                                            |
| New-StatusCakeHelperPublicReportingPage      	| Removed no longer supported by API                                            |
| Remove-StatusCakeHelperPublicReportingPage   	| Removed no longer supported by API                                           	|
| Set-StatusCakeHelperPublicReportingPage     	| Removed no longer supported by API                                        	|
|                                               |                                                                               |
| Add-StatusCakeHelperTestNodeLocation        	| Removed                                                                    	|
| Add-StatusCakeHelperTestStatusCode        	| Removed                                                                      	|
| Add-StatusCakeHelperTestTag                 	| Removed                                                                     	|
| Remove-StatusCakeHelperTestNodeLocation     	| Removed                                                                    	|
| Remove-StatusCakeHelperTestStatusCode     	| Removed                                                                      	|
| Remove-StatusCakeHelperTestTag               	| Removed                                                                     	|
|                                               |                                                                               |
| Copy-StatusCakeHelperTest                     | Renamed to Copy-StatusCakeHelperUptimeTest                                    |
| Get-StatusCakeHelperPausedTest                | Renamed to Get-StatusCakeHelperUptimePausedTest                               |
| Get-StatusCakeHelperTest                      | Renamed to Get-StatusCakeHelperUptimeTest                                     |
| New-StatusCakeHelperTest                      | Renamed to New-StatusCakeHelperUptimeTest                                     |
| Remove-StatusCakeHelperTest                   | Renamed to Remove-StatusCakeHelperUptimeTest                                  |
| Resume-StatusCakeHelperTest                   | Renamed to Resume-StatusCakeHelperUptimeTest                                  |
| Resume-StatusCakeHelperTest                   | Renamed to Resume-StatusCakeHelperUptimeTest                                  |
| Suspend-StatusCakeHelperTest                  | Renamed to Suspend-StatusCakeHelperUptimeTest                                 |
|                                               |                                                                               |

## Release 4.0

This release makes changes to the authentication parameters used by the functions, standardizes the outputs, brings more of the code in-line with Powershell conventions and implements a number of PSScript Analyzer recommendations.

## All sections

*  Username and ApiKey parameters have been replaced by APICredential parameter
*  baseURL parameter has been removed

StatusCake-Helpers/Public/New-StatusCakeHelperSSLTest.ps1

<#
.Synopsis
   Create a StatusCake SSL Test
.EXAMPLE
   New-StatusCakeHelperSSLTest -Username "Username" -ApiKey "APIKEY" -Domain "https://www.example.com" -checkrate 3600
.INPUTS
    baseSSLTestURL - Base URL endpoint of the statuscake ContactGroup API
    Username - Username associated with the API key
    ApiKey - APIKey to access the StatusCake API
    
    Domain - URL to check SSL certificate, must begin with https://
    CheckRate - Checkrate in seconds.
    Contact_Groups - Array containing contact IDs to alert.
    Alert_At - Number of days before expiration when reminders will be sent. Defaults to reminders at 60, 30 and 7 days. Must be 3 numeric values.
    Alert_expiry - 	Set to true to disable expiration alerts.
    Alert_reminder - Set to true to disable reminder alerts.
    Alert_broken - Set to true to disable broken alerts

.FUNCTIONALITY
   Creates a new StatusCake SSL Test using the supplied parameters.
#>
function New-StatusCakeHelperSSLTest
{
    [CmdletBinding(PositionalBinding=$false,SupportsShouldProcess=$true)]    
    Param(
        $baseSSLTestURL = "https://app.statuscake.com/API/SSL/Update",
        [Parameter(Mandatory=$true)]        
        $Username,
        [Parameter(Mandatory=$true)]        
        $ApiKey,

        [Parameter(Mandatory=$true)] 
        [ValidatePattern('^((https):\/\/)([a-zA-Z0-9\-]+(\.[a-zA-Z]+)+.*)$|^(?!^.*,$)')]                  
        $domain,

        #Contact_Groups must be supplied
        [ValidateScript({$_ -match '^[\d]+$'})] 
        [object]$contact_groups="0",

        [Parameter(Mandatory=$true)] 
        [ValidateSet("300","600","1800","3600","86400","2073600")]
        $checkrate,
        
        [ValidateScript({$_ -match '^[\d]+$'})]        
        [object]$alert_at=@("7","30","60"),

        [ValidateRange(0,1)] 
        $alert_expiry=1,

        [ValidateRange(0,1)] 
        $alert_reminder=1,

        [ValidateRange(0,1)] 
        $alert_broken=1

    )
    $authenticationHeader = @{"Username"="$username";"API"="$ApiKey"}
 
    if($Alert_At.count -ne 3)
    {
        Write-Error "Only three values must be specified for Alert_At parameter"
        Return
    }

    if( $pscmdlet.ShouldProcess("StatusCake API", "Retrieve StatusCake SSL Checks") )
    {
        $sslTest = Get-StatusCakeHelperSSLTest -Username $username -apikey $ApiKey -domain $domain
        if($sslTest)
        {
            Write-Error "SSL Check with specified domain already exists [$domain] [$($sslTest.id)]"
            Return $null 
        }
    }    

    $psParams = @{}
    $ParameterList = (Get-Command -Name $MyInvocation.InvocationName).Parameters
    $ParamsToIgnore = @("baseSSLTestURL","Username","ApiKey")
    foreach ($key in $ParameterList.keys)
    {
        $var = Get-Variable -Name $key -ErrorAction SilentlyContinue;
        
        if($ParamsToIgnore -contains $var.Name)
        {
            continue
        }
        elseif($var.value -or $var.value -eq 0)
        {   
            $psParams.Add($var.name,$var.value)                  
        }
    }

    $statusCakeAPIParams = $psParams | ConvertTo-StatusCakeHelperAPIParams

    $putRequestParams = @{
        uri = $baseSSLTestURL
        Headers = $authenticationHeader
        UseBasicParsing = $true
        method = "Put"
        ContentType = "application/x-www-form-urlencoded"
        body = $statusCakeAPIParams 
    }

    if( $pscmdlet.ShouldProcess("StatusCake API", "Add StatusCake SSL Test") )
    {
        $jsonResponse = Invoke-WebRequest @putRequestParams
        $response = $jsonResponse | ConvertFrom-Json
        if($response.Success -ne "True")
        {
            Write-Error "$($response.Message) [$($response.Issues)]"
            Return $null
        }         
        Return $response
    }

}